File: crates/agentpin-cli/src/keygen.rs
use std::fs;
use std::path::Path;

use clap::Args;

use agentpin::crypto;
use agentpin::jwk::pem_to_jwk;

#[derive(Args)]
pub struct KeygenArgs {
    /// Domain this key is associated with
    #[arg(long)]
    pub domain: String,

    /// Key identifier (e.g., "example-2026-01")
    #[arg(long)]
    pub kid: String,

    /// Output directory for key files
    #[arg(long, default_value = ".")]
    pub output_dir: String,

    /// Output format
    #[arg(long, default_value = "both", value_parser = ["jwk", "pem", "both"])]
    pub format: String,
}

pub fn run(args: KeygenArgs) -> anyhow::Result<()> {
    let kp = crypto::generate_key_pair()?;

    let out_dir = Path::new(&args.output_dir);
    fs::create_dir_all(out_dir)?;

    let wrote_private_pem;
    let wrote_public_pem;
    let wrote_public_jwk;

    match args.format.as_str() {
        "pem" => {
            let priv_path = out_dir.join(format!("{}.private.pem", args.kid));
            fs::write(&priv_path, &kp.private_key_pem)?;
            wrote_private_pem = Some(priv_path);

            let pub_path = out_dir.join(format!("{}.public.pem", args.kid));
            fs::write(&pub_path, &kp.public_key_pem)?;
            wrote_public_pem = Some(pub_path);
            wrote_public_jwk = None;
        }
        "jwk" => {
            let priv_path = out_dir.join(format!("{}.private.pem", args.kid));
            fs::write(&priv_path, &kp.private_key_pem)?;
            wrote_private_pem = Some(priv_path);

            let jwk = pem_to_jwk(&kp.public_key_pem, &args.kid)?;
            let jwk_json = serde_json::to_string_pretty(&jwk)?;
            let jwk_path = out_dir.join(format!("{}.public.jwk.json", args.kid));
            fs::write(&jwk_path, &jwk_json)?;
            wrote_public_jwk = Some(jwk_path);
            wrote_public_pem = None;
        }
        _ => {
            let priv_path = out_dir.join(format!("{}.private.pem", args.kid));
            fs::write(&priv_path, &kp.private_key_pem)?;
            wrote_private_pem = Some(priv_path);

            let pub_path = out_dir.join(format!("{}.public.pem", args.kid));
            fs::write(&pub_path, &kp.public_key_pem)?;
            wrote_public_pem = Some(pub_path);

            let jwk = pem_to_jwk(&kp.public_key_pem, &args.kid)?;
            let jwk_json = serde_json::to_string_pretty(&jwk)?;
            let jwk_path = out_dir.join(format!("{}.public.jwk.json", args.kid));
            fs::write(&jwk_path, &jwk_json)?;
            wrote_public_jwk = Some(jwk_path);
        }
    }

    eprintln!(
        "Generated ECDSA P-256 keypair for domain '{}' (kid: '{}')",
        args.domain, args.kid
    );
    if let Some(p) = wrote_private_pem {
        eprintln!("  Private key: {}", p.display());
    }
    if let Some(p) = wrote_public_pem {
        eprintln!("  Public key (PEM): {}", p.display());
    }
    if let Some(p) = wrote_public_jwk {
        eprintln!("  Public key (JWK): {}", p.display());
    }

    Ok(())
}

File: go/cmd/agentpin/keygen.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"

	"github.com/ThirdKeyAi/agentpin/go/pkg/crypto"
	"github.com/ThirdKeyAi/agentpin/go/pkg/jwk"
)

func runKeygen(args []string) error {
	fs := flag.NewFlagSet("keygen", flag.ContinueOnError)
	domain := fs.String("domain", "", "Domain this key is associated with (required)")
	kid := fs.String("kid", "", "Key identifier, e.g. 'example-2026-01' (required)")
	outputDir := fs.String("output-dir", ".", "Output directory for key files")
	format := fs.String("format", "both", "Output format: jwk, pem, or both")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *domain == "" || *kid == "" {
		return fmt.Errorf("--domain and --kid are required")
	}

	kp, err := crypto.GenerateKeyPair()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(*outputDir, 0o755); err != nil {
		return err
	}

	privPath := filepath.Join(*outputDir, *kid+".private.pem")
	if err := os.WriteFile(privPath, []byte(kp.PrivateKeyPEM), 0o600); err != nil {
		return err
	}
	fmt.Fprintln(os.Stderr, "Generated ECDSA P-256 keypair for domain", *domain, "(kid:", *kid+")")
	fmt.Fprintln(os.Stderr, "  Private key:", privPath)

	if *format == "pem" || *format == "both" {
		pubPath := filepath.Join(*outputDir, *kid+".public.pem")
		if err := os.WriteFile(pubPath, []byte(kp.PublicKeyPEM), 0o644); err != nil {
			return err
		}
		fmt.Fprintln(os.Stderr, "  Public key (PEM):", pubPath)
	}
	if *format == "jwk" || *format == "both" {
		j, err := jwk.PEMToJWK(kp.PublicKeyPEM, *kid)
		if err != nil {
			return err
		}
		jb, err := json.MarshalIndent(j, "", "  ")
		if err != nil {
			return err
		}
		jwkPath := filepath.Join(*outputDir, *kid+".public.jwk.json")
		if err := os.WriteFile(jwkPath, jb, 0o644); err != nil {
			return err
		}
		fmt.Fprintln(os.Stderr, "  Public key (JWK):", jwkPath)
	}

	return nil
}
